superduperdb/ext/llm/vllm.py

```diff
@@ -0,0 +1,169 @@
+
+    __doc__ = __doc__.format(parent_doc=BaseLLMAPI.__doc__)
+
+    def _generate(self, prompt: str, **kwargs) -> str:
+        """
+        Batch generate text from a prompt.
+        """
+        post_data = self.build_post_data(prompt, **kwargs)
+        response = requests.post(self.api_url, json=post_data)
+        return response.json()["text"][0]
+
+    def _batch_generate(self, prompts: List[str], **kwargs: Any) -> List[str]:
+        """
+        Use asyncio to batch generate text from a list of prompts.
+        """
+        return asyncio.run(self._async_batch_generate(prompts, **kwargs))
+
+    async def _async_generate(self, session, semaphore, prompt: str, **kwargs) -> str:
+        post_data = self.build_post_data(prompt, **kwargs)
+        async with semaphore:
+            try:
+                async with session.post(self.api_url, json=post_data) as response:
+                    response_json = await response.json()
+                    return response_json["text"][0]
+            except aiohttp.ClientError as e:
+                logging.error(f"HTTP request failed: {e}. Prompt: {prompt}")
+                return ""
+
+    async def _async_batch_generate(self, prompts: List[str], **kwargs) -> List[str]:
+        async with aiohttp.ClientSession() as session:
+            semaphore = asyncio.Semaphore(self.max_batch_size or len(prompts))
+            tasks = [
+                self._async_generate(session, semaphore, prompt, **kwargs)
+                for prompt in prompts
+            ]
+            return await asyncio.gather(*tasks)
+
+    def build_post_data(self, prompt: str, **kwargs: dict[str, Any]) -> dict[str, Any]:
+        total_kwargs = {}
+        for key, value in {**self.inference_kwargs, **kwargs}.items():
+            if key in VLLM_INFERENCE_PARAMETERS_LIST:
+                total_kwargs[key] = value
+        return {"prompt": prompt, **total_kwargs}
+
+
+@dc.dataclass
+class VllmModel(BaseLLMModel):
+    """
+    Load a large language model from VLLM.
+
+    :param model_name: The name of the model to use.
+    :param trust_remote_code: Whether to trust remote code.
+    :param dtype: The data type to use.
+    {parent_doc}
+    """
+
+    __doc__ = __doc__.format(parent_doc=BaseLLMModel.__doc__)
+
+    tensor_parallel_size: int = 1
+    trust_remote_code: bool = True
+    vllm_kwargs: dict = dc.field(default_factory=dict)
+
+    def __post_init__(self):
+        self.on_ray = self.on_ray or bool(self.ray_address)
+        super().__post_init__()
+
+    def init(self):
+        try:
+            from vllm import LLM
+        except ImportError:
+            raise Exception("You must install vllm with command 'pip install vllm'")
+
+        if self.on_ray:
+            try:
+                import ray
+            except ImportError:
+                raise Exception("You must install vllm with command 'pip install ray'")
+
+            runtime_env = {"pip": ["vllm"]}
+            if not ray.is_initialized():
```

Comment:
can we somehow utilise RayCompute from backends/ray/compute.py ?

Reply:
However, perhaps when SuperDuperDB uses Ray as its computation backend, it could by default use the same Ray instance, but it must be equipped with a GPU.